src/action.js

```diff
@@ -0,0 +1,90 @@
+class Action {
+  constructor(gameId, player, timestamp = Date.now()) {
+    this.gameId = gameId;
+    this.player = player;
+    this.timestamp = timestamp;
+  }
+}
+
+class TokenAction extends Action {
+  constructor(tokens, discardedTokens, ...args) {
+    super(...args);
+    this.tokens = tokens;
+    this.discardedTokens = discardedTokens;
+
+    if (this.tokens.total() > 3)
+      throw new Error("No more than 3 tokens can be drawn");
+    if (this.tokens.total() === 3 && this.tokens.max() !== 1)
+      throw new Error("Only 1 per color can be drawn when drawing 3 tokens");
+  }
+
+  apply(game) {
+    if (game.hasTokens(this.tokens)) {
+      // TODO: Check for remaining pool for 2 token draws
+      game.removeTokens(this.tokens);
+      this.player.addTokens(this.tokens);
+      console.log(`${this.player.name} drew ${this.tokens} tokens`);
+    }
+    if (!this.discardedTokens.isEmpty()) {
+      this.player.removeTokens(this.discardedTokens);
+      game.addTokens(this.discardedTokens);
+      console.log(
+        `${this.player.name} discarded ${this.discardedTokens} tokens`
+      );
+    }
+  }
+}
+
+class BuyAction extends Action {
+  constructor(id, ...args) {
+    super(...args);
+    this.id = id;
+    // card = card;
+  }
+
+  apply(game) {
+    const card = game.getCardById(id);
```

Comment:
```suggestion
    const card = game.getCardById(this.id);
```
Isis supposed to be like this?